File: machinePlayer.py
import node
import numpy as np
import random
import player


class machinePlayer(player.Player):

    def __init__(self):
        self.__root = node.Node(p=None, n=None, v=0)
        self.__temp = self.__root
        self.__size = 1
        self.__allChoices = [(0, 0), (0, 1), (0, 2), (1, 0),
                             (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
        ####### board ########
        #(0, 0) (1, 0) (2, 0)#
        #(0, 1) (1, 1) (2, 1)#
        #(0, 2) (1, 2) (2, 2)#
        ######################
        self.__path = []  # [((x,y),player), ((x,y),player), ...]

    def root(self):
        return self.__root

    def isRoot(self, n):
        if n.getParent() == None:
            return True
        return False

    def isExternal(self, n):
        if n.getChildrenList() == []:
            return True
        return False

    def isInternal(self, n):
        if n.getChildrenList() == []:
            return False
        return True

    def size(self):
        return self.__size

    def updateValue(self, n, v):
        n.setValue((n.getValue()[0]+v[0], n.getValue()[1]+v[1]))
        return n

    def updatePath(self, pos, playerName):
        self.__allChoices.remove(pos)
        self.__path.append((pos, playerName))

    def moveWithOpponent(self, opponentName, opponentMovePos):
        if self.isExternal(self.__temp):
            self.expand()
        children_list = self.__temp.getChildrenList()
        updated = False
        for each in children_list:
            if each.getName() == opponentMovePos:
                self.__temp = each
                self.updatePath(pos=opponentMovePos, playerName=opponentName)
                updated = True
                break
        if not updated:
            print("Faild to update!!!")
            quit()

    def clearPath(self):
        self.__path.clear()
        self.__allChoices = [(0, 0), (0, 1), (0, 2), (1, 0),
                             (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]

    def doMinMax(self, childrenList):
        minMax = []  # minMax: [[max score, 0th node],[max score, 1st node],...]
        for i in range(len(childrenList)):
            # child i has been explored before.
            if self.isInternal(childrenList[i]):
                ll = childrenList[i].getChildrenList()
                maxScore = -float('inf')
                this = 0
                for each in ll:
                    # The denominator is not 0, which means child i's child(each) has been visited.
                    if each.getValue()[1] != 0:
                        this = each.getValue()[0]/each.getValue()[1]
                    # The denominator is 0, which means child i's child(each) has never been visited.
                    else:
                        # This can be changed if you want to explore nodes that hasn't been visited.
                        # this = 0
                        # Strategy 2: tend to explore new paths
                        this = -float('inf')
                    if this > maxScore:
                        maxScore = this
                minMax.append([maxScore, i])
            else:
                # "getValue()[1]!=0" and "isExternal" mean child i is the last step,
                # and if child i is a "winning" step, getValue()[0] must > 0, so...
                if childrenList[i].getValue()[0] > 0 and childrenList[i].getValue()[1] != 0:
                    return [[-float('inf'), i], False]
                # or if child i is a "losing" step...
                elif childrenList[i].getValue()[0] < 0 and childrenList[i].getValue()[1] != 0:
                    minMax.append([float('inf'), i])
                # or if child i is a "drawing" step (must be the 9th step)...
                elif childrenList[i].getValue()[0] == 0 and childrenList[i].getValue()[1] != 0:
                    minMax.append([0, i])
                # "getValue()[1]==0" and "isExternal" mean child i has never been visited...
                elif childrenList[i].getValue()[1] == 0:
                    # Strategy 1: tend to explore new paths
                    minMax.append([-float('inf'), i])
                    # Strategy 2: tend to follow the previous-winning path
                    # minMax.append([0, i])
        minMax.sort()   # small to large
        minMaxNp = np.array(minMax)[:, 0]
        allTheSame = np.all(minMaxNp == minMaxNp[0])
        return [minMax[0], allTheSame]

    ################ CORE!!!!##################
    def select(self, playerName):
        if self.isExternal(self.__temp):
            self.expand()
            child_list = self.__temp.getChildrenList()
            self.__temp = child_list[random.randint(
                0, len(child_list)-1)]    # Rollout Policy
        else:
            child_list = self.__temp.getChildrenList()
            [minMax, allTheSame] = self.doMinMax(child_list)
            if allTheSame:
                self.__temp = child_list[random.randint(
                    0, len(child_list)-1)]    # Rollout Policy
            else:
                self.__temp = child_list[minMax[1]]
        # Because we use position as the name of a node at very first, so...
        position = self.__temp.getName()
        self.updatePath(position, playerName)
        return position

    def expand(self):
        random.shuffle(self.__allChoices)
        for each in self.__allChoices:
            # Use position as the name of the node expanded:
            self.__temp.setChild(node.Node(p=self.__temp, n=each, v=(0, 0)))

    # Mind: In this game, last mover is impossible to be a loser.
    def doBackpropagation(self, winLossTie):
        i = -1
        lastMover = self.__path[i][1]
        if winLossTie == 1:  # last mover won
            while not self.isRoot(self.__temp):
                if self.__path[i][1] == lastMover:
                    self.updateValue(n=self.__temp, v=(1, 1))
                else:
                    self.updateValue(n=self.__temp, v=(-1, 1))
                self.__temp = self.__temp.getParent()
                i -= 1
        else:  # last mover made it tie
            while not self.isRoot(self.__temp):
                self.updateValue(n=self.__temp, v=(0, 1))
                self.__temp = self.__temp.getParent()
                i -= 1

    def postOrder(self, n):
        if n.getChildrenList() != []:
            print(str(n.getName())+"\t" +
                  str(n.getValue())+"\t" +
                  str(n.getParent()).split(" ")[-1][-9:-1]+"\t" +
                  str(len(n.getChildrenList()))+"  " +
                  str(n.printChildrenList()))
        for each in n.getChildrenList():
            self.postOrder(each)

File: webpage/modules/machinePlayer.js
import { Node } from './node.js';
export class MachinePlayer {
    constructor() {
        this._root = new Node(null, "ROOT", [null, Infinity]);
        this._temp = this._root;
        this._allChoices = [[0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1], [2, 2]];
        this._path = [];
    }
    isExternal(n) {
        return n.childrenList.length == 0;
    }
    updatePath(pos, playerName) {
        this._allChoices = this._allChoices.filter(each => {
            return (each[0] != pos[0] || each[1] != pos[1]);
        });
        this._path.push([pos, playerName]);
    }
    moveWithOpponent(opponentName, opponentMovePos) {
        if (this.isExternal(this._temp))
            this.expand();
        let tempChildrenList = this._temp.childrenList;
        for (let i = 0; i < tempChildrenList.length; i++) {
            if (tempChildrenList[i].name[0] == opponentMovePos[0] && tempChildrenList[i].name[1] == opponentMovePos[1]) {
                this._temp = tempChildrenList[i];
                this.updatePath(opponentMovePos, opponentName);
                break;
            }
        }
    }
    clearPath() {
        this._path = [];
        this._allChoices = [[0, 0], [0, 1], [0, 2], [1, 0], [1, 1], [1, 2], [2, 0], [2, 1], [2, 2]];
    }
    select(playerName) {
        if (this.isExternal(this._temp))
            this.expand();
        let tempChildrenList = this._temp.childrenList;
        let pos = null;
        for (let i = 0; i < tempChildrenList.length; i++) {
            if (tempChildrenList[i].name == this._temp.value[0]) {
                this._temp = tempChildrenList[i];
                pos = this._temp.name;
                this.updatePath(pos, playerName);
                break;
            }
        }
        return pos;
    }
    shuffle(array) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }
    expand() {
        this.shuffle(this._allChoices);
        this._allChoices.forEach(each => {
            this._temp.appendChild(new Node(this._temp, each, [null, Infinity]));
        });
        this.backPropagate("2");
    }
    backPropagate(state) {
        // states: "2" means that this' for expanding,
        //         "1" means that the first mover won,
        //         "-1" means that the first mover lost,
        //         "0" means that this game went tie.
        let probe;
        probe = state == "2" ? this._temp.childrenList[0] : this._temp;
        let depth = 0;
        while (probe.parent != null) {
            probe = probe.parent;
            probe.value = [null, null];
            depth++;
        }
        if (state == "1")
            this._temp.value = [null, 100 / depth];
        else if (state == "-1")
            this._temp.value = [null, -100 / depth];
        else if (state == "0")
            this._temp.value = [null, 0];
        this.minimax(probe, true);
        if (state != "2")
            this._temp = probe;
    }
    hasNullValue(aListOfNodes) {
        let hasNull = false;
        let items = [];
        for (let i = 0; i < aListOfNodes.length; i++) {
            if (aListOfNodes[i].value[0] == null && aListOfNodes[i].value[1] == null) {
                hasNull = true;
                items.push(i);
            }
        }
        return [hasNull, items];
    }
    minimax(aTree, isMaximizer) {
        let needRecursion = this.hasNullValue(aTree.childrenList);
        if (needRecursion[0]) {
            needRecursion[1].forEach(each => {
                this.minimax(aTree.childrenList[each], !isMaximizer);
            });
        }
        let cInfo = [];
        aTree.childrenList.forEach(each => {
            cInfo.push([each.name, each.value[1]]);
        });
        function ascendingSort(a, b) {
            if (a[1] == b[1])
                return 0;
            else
                return a[1] < b[1] ? -1 : 1;
        }
        function descendingSort(a, b) {
            if (a[1] == b[1])
                return 0;
            else
                return (a[1] < b[1]) ? 1 : -1;
        }
        let v;
        v = isMaximizer ? cInfo.sort(descendingSort)[0] : cInfo.sort(ascendingSort)[0];
        aTree.value = v;
    }
}
